pulp_ansible/app/migrations/0036_update_repository_content.py

```diff
@@ -9,56 +9,77 @@ def create_repository_content_for_deprecation(apps, schema_editor):
+                RepositoryContent.objects.bulk_create(repo_content)
+                repo_content.clear()
+    RepositoryContent.objects.bulk_create(repo_content)
+    repo_content.clear()
+    AnsibleCollectionDeprecated.objects.filter(repository_id__isnull=False).delete()
 
+    # Update Repository content counts
     for repo_version_id, total in count.items():
         count_obj = RepositoryVersionContentDetails(
-            content_type=deprecation.pulp_type,
+            content_type="collection_deprecation",
             repository_version_id=repo_version_id,
             count=total,
             count_type="A", # Added
         )
         counts_list.append(count_obj)
 
         count_obj = RepositoryVersionContentDetails(
-            content_type=deprecation.pulp_type,
+            content_type="collection_deprecation",
             repository_version_id=repo_version_id,
             count=total,
             count_type="P", # Present
         )
         counts_list.append(count_obj)
+        if len(counts_list) >= 1024:
+            RepositoryVersionContentDetails.objects.bulk_create(counts_list)
+            counts_list.clear()
 
-    RepositoryContent.objects.using(db_alias).bulk_create(repo_content)
-    RepositoryVersionContentDetails.objects.using(db_alias).bulk_create(counts_list)
-    AnsibleCollectionDeprecated.objects.using(db_alias).filter(version_memberships__isnull=True).delete()
+    RepositoryVersionContentDetails.objects.bulk_create(counts_list)
 
 
 def create_deprecation_from_repository_content(apps, schema_editor):
     """ Create AnsibleCollectionDeprecated from RepositoryContent."""
     AnsibleCollectionDeprecated = apps.get_model('ansible', 'AnsibleCollectionDeprecated')
     Collection = apps.get_model('ansible', 'Collection')
+
     for deprecation in AnsibleCollectionDeprecated.objects.all():
         col = Collection.objects.get(namespace=deprecation.namespace, name=deprecation.name)
         for repo_content in deprecation.version_memberships.all():
-            deprecation.collection_id = col.pk
-            deprecation.repository_id = repo_content.repository_id
-            deprecation.version_added_id = repo_content.version_added_id
-            deprecation.save()
+            AnsibleCollectionDeprecated.objects.create(
+                name=deprecation.name,
+                namespace=deprecation.namespace,
+                collection_id=col.pk,
+                repository_id=repo_content.repository_id,
+                version_added_id = repo_content.version_added_id,
+            )
+
+    AnsibleCollectionDeprecated.objects.filter(repository_id__isnull=True).delete()
 
 
 class Migration(migrations.Migration):
 
+    # You cannot change data and modify the a table in the same migration.
+    # Since this is an already released migration, we cannot split this one in two.
+    # So the only resort is running the steps of this migration without transactions.
+    atomic = False
```

Comment:
@fao89 @bmbouter I'd rather avoid this, but i see no option.

Reply:
I'm not sure of the answer here so feel free to educate me. It seems the issue of one transaction with data changes and SQL alter changes is the crux of the problem. If that's the case and they want to be separated how abut we move all data changes to one migration and leave all SQL alter changes in the other? I think I'm still clinging to the (maybe incorrect) belief that more of this diff should be happening in migration 35. You would know better than I, wdyt?

Reply:
If we were to introduce these migrations first, i'd agree. But as the matter of things is, a user may have migrated beyond 35 already and failing to migrate 36. At that point we need to fix 36 for him. And i believe we cannot introduce new migrations in the chain. That is why i think, we are stuck fixing it _in_ 36.

Reply:
I think I only now understand what you were trying to tell me before. What you are saying makes sense, but I've not heard of any users actually ending up in that state. I do worry about the risk to all users with this being atomic to serve a perhaps theoretical user base. We're picking from the lesser of two evils here so I won't claim fixing migration 35 is a big win, but I do think it's better. @mdellweg what do you think about this perspective?

Reply:
I am the theoretician here. And my biggest concern is that we might redefine the scope of already published migrations.
Also when stuck in that position between migrations 35 and 36 the only way to repair (apart from a complete rollback from backup) would be to checkout the old codebase, revert 35, then install the new codebase again and migrate forward. So the rule that i think applies here (for exactly that crazy dance) is "we don't change existing migrations".
OTOH, i see that migrations 35 and 36 always appear in pairs, and when applied successfully, the result is good enough before _and_ after the change. So the only way to get in the bad situation will be a single failed upgrade. And there you just need to be extra careful to complete all shipped migrations with the same codebase.
Having expressed my concerns loudly here, I yield and will have a go at the combined fix.